### PR TITLE
Add : Heatmap Layer

### PR DESCRIPTION
**Pull request by 1chandu**
For #2 

#### Background
- Bump luma.gl to 7.2.0-alpha.5
- Add new  `HeatmapLayer`
- Add demo to layer-browser and a new example in test/apps
- Add doc page and demo.

Added unit tests for utility methods will be adding more unit tests for Layer in next PR.

heatmap-pr2

#### Change List
- Bump luma.gl to 7.2.0-alpha.5
- HeatmapLayer : Add new layer.

### CODE REVIEW

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import {experimental, createIterable} from '@deck.gl/core';
+const {count} = experimental;
+
+// Parse input data to build positions, wights arrays.
+export function parseData(data, getPosition, getWeight) {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: Use attribute manager?

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  }
+  return false;
+}
+
+// For given rectangle bounds generates two triangles vertices that coverit completely
+export function getTriangleVertices(opts = {}) {
+  const {xMin = 0, yMin = 0, xMax = 1, yMax = 1, addZ = false} = opts;
+
+  if (addZ) {
+    return new Float32Array([
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: This is not necessary - you can provide a `size:2` buffer to a vec3 attribute.

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  const yCenter = (yMax + yMin) / 2;
+
+  return [
+    [xCenter - newWidth / 2, yCenter - newHeight / 2],
+    [xCenter + newWidth / 2, yCenter + newHeight / 2]
+  ];
+}
+
+// Scales texture coordiante range to a sub rectangel
+export function scaleTextureCoordiantes(originalRect, subRect) {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: - Spelling
- `getTextureCoordinates`?

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+      }),
+      triTexCoordBuffer: new Buffer(gl, {
+        data: getTriangleVertices(),
+        accessor: {size: 2}
+      })
+    };
+  }
+
+  shouldUpdateState({changeFlags}) {
+    // Need to be updated when viewports
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: Comment incomplete

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    return changeFlags.somethingChanged;
+  }
+
+  // returns visible world bounds [[minLong, minLat], [maxLong, maxLat]
+  getVisibleWorldBounds() {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: Use underscore prefix for private methods

File: modules/aggregation-layers/src/heatmap-layer/triangle-layer.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        geometry: new Geometry({
+          drawMode: GL.TRIANGLES,
+          vertexCount: count
+        }),
+        shaderCache: this.context.shaderCache
+      })
+    );
+  }
+
+  calculatePositions(attribute) {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/triangle-layer.js]: This is automatically handled by the attribute manager.

**Comment by Pessimistress** [comment]
Do not check in the sample data.

File: modules/aggregation-layers/src/heatmap-layer/triangle-layer.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import vs from './triangle-layer-vertex.glsl';
+import fs from './triangle-layer-fragment.glsl';
+
+const defaultProps = {
+  // data {positions(Buffer), texCoords(Buffer)}
+  count: 0, // number of triangles to be rendered
+  texture: null
+};
+
+export default class TriangleLayer extends Layer {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/triangle-layer.js]: Ideally you don't need another layer to render this. You can easily create the model inside `HeatmapLayer` and update its attributes manually, as you are already doing.

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js (hunk 1 above)
Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: Thanks for the help @Pessimistress , updated with attributeManager integration.

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js (hunk 3 above)
Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: Done

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js (hunk 5 above)
Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: Done.

File: modules/aggregation-layers/src/heatmap-layer/triangle-layer.js (hunk 7 above)
Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/triangle-layer.js]: Intention is to use this layer in future for Contours.

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js (hunk 2 above)
Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/heatmap-layer-utils.js]: Keeping it for now, to enable `elevation`, there was one user requesting it.

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+
+    if (changeFlags.dataChanged || changeFlags.boundsChanged || changeFlags.uniformsChanged) {
+      this._updateWeightmap();
+    }
+
+    if (props.colorRange !== oldProps.colorRange) {
+      this._updateColorTexture(opts);
+    }
+
+    // TODO: need to be fixed, seem to be rendring wrong part of the texture
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: This comment no longer applies?

File: modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+      id: this.props.id,
+      stats: this.context.stats
+    });
+  }
+
+  _getChangeFlags(opts) {
+    const {oldProps, props} = opts;
+    const changeFlags = {};
+    if (this._isDataChanged(opts)) {
+      changeFlags.dataChanged = true;
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: Try not to overwrite existing flags. This may have unexpected consequences.

Comment (1chandu) [reply modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js]: Only changing local object, `changeFlags`.

**Pull request merged**